pkg/apiclient/client_factory.go

```diff
@@ -70,13 +72,13 @@ type Client struct {
 	Ask question.AskProvider
 }
 
-func NewClientFactory(httpClient *http.Client, host string, apiKey string, spaceNameOrID string, ask question.AskProvider) (ClientFactory, error) {
+func NewClientFactory(httpClient *http.Client, host string, apiKey string, accessToken string, spaceNameOrID string, ask question.AskProvider) (ClientFactory, error) {
```

Comment:
I am not familiar with this code base, but to me, it feels weird to have both `apiKey` and `accessToken` in the initialiser function, aren't they mutually exclusive?

Reply:
Technically they aren't mutually exclusive, if you supply both it will use both and the api key will win out on the Server. But yes for practical purposes you would only really use one or the other. In the client library we introduced a struct for this that's used in some function overloads (leaving the original ones in place to maintain backward compat), we could do the same here potentially. I'll have a look and see what the changes look like, it might cause the PR to get pretty large with changes to tests but that could be ok if it ends up more obvious and maintainable. Thanks.

Reply:
One other thought here is backward compatibility on these functions. I had made an assumption that we don't need to ensure this as we don't publish the packages for consumption elsewhere, I can't find the cli with a search https://pkg.go.dev/search?q=octopusdeploy, only our client and few other third-party packages. @benPearce1 do you know if this assumption is correct?

Reply:
The only other place the `go-octopusdeploy` library is used is in our terrraform provider. I think it might be used by individuals as well, but not in other "products"

Reply:
The change in `go-octopusdeploy` are backward compatible, we left the original functions in there. I'm more thinking of the cli itself, I don't think it gets published as a consumable go module, only as a binary, is that right?

Reply:
Sorry, misunderstood. That is correct.

Reply:
Thanks, I'll take a look at what the changes might look like to have a single credentials parameter instead of two like we do in `go-octopusdeploy`, might be worth it if the result is cleaner and easier to follow.

Reply:
The change that we might prefer might look something like this in the client library

```
type ICredential interface {
	GetHeaderValue() (string, string)
}

type ApiKey struct {
	Value string
	ICredential
}

func (apikey *ApiKey) GetHeaderValue() (string, string) {
	return constants.ClientAPIKeyHTTPHeader, apikey.Value
}

func NewApiKey(apiKey string) (*ApiKey, error) {
	return &ApiKey{
		Value: apiKey,
	}, nil
}

type AccessToken struct {
	Value string
	ICredential
}

func (accessToken *AccessToken) GetHeaderValue() (string, string) {
	return "Authorization", fmt.Sprintf("Bearer %s", accessToken.Value)
}

func NewAccessToken(accessToken string) (*AccessToken, error) {
	return &AccessToken{
		Value: accessToken,
	}, nil
}

func getHeaders(apiCredentials ICredential, requestingTool string) map[string]string {
	headers := map[string]string{
		"User-Agent": api.GetUserAgentString(requestingTool),
	}

	headerKey, headerValue := apiCredentials.GetHeaderValue()
	headers[headerKey] = headerValue

	return headers
}
```

Reply:
That is certainly nicer than what we have at the moment, I'm very new to go and so went with the simplest option I could do at the time but this is a nicer approach. If we were to do this change in the client and then roll it to the cli, given we've shipped the new minor version of the client in its current form would it constitute a breaking change?

Reply:
In theory, the previous change was also a breaking change due to the new parameter.

Reply:
We only use it in new functions and kept the old ones around that passed through the correct values, so I didn't consider it to be a breaking change?

Reply:
Cool, didn't see that. Then yes, this proposal would be a breaking change.

Reply:
We only shipped this a couple of days ago, would we need to consider a new major version with this change or can we just push a new minor? It would be nice to avoid a major version if possible.

Reply:
Actually it was only yesterday that we shipped it

Reply:
These go projects aren't using strict semver, a minor version bump would be sufficient

Reply:
Ok great, I'll make the change and might send it your way for a review if thats ok

Reply:
No problem

Reply:
Changes to the client have been made in https://github.com/OctopusDeploy/go-octopusdeploy/pull/202 and the equivalent changes made in this PR. It won't build just yet as it'll require the client changes to be released first. @benPearce1 do you mind also eyeballing the changes here to see if the usage feels good now please? If so I'll merge the client PR and get this one updated.